### ISSUE TITLE
Sparse-dense matmul failing in 3.5

### ISSUE THREAD

**Issue by villekf** [arrayfire/arrayfire#1866]
Sparse-dense matrix multiplication is failing in 3.5.0 both on Windows and Ubuntu when using either CUDA or OpenCL backend (haven't tested CPU). The same code works fine when using 3.4.2. On Windows the error message is:

`Unexpected  Standard exception from MEX file.
What() is:ArrayFire Exception (Unknown error:999):

In function class af::array __cdecl af::matmul(const class af::array &,const class af::array &,const
af_mat_prop,const af_mat_prop)
In file src\api\cpp\blas.cpp:20'

On Ubuntu the computations freeze completely with no error message. On both cases the official installer was used as well as CUDA 8.0. Normal dense-dense matmul works fine.

**Comment by pavanky** [comment]
@villekf What GPU are you using?

**Comment by villekf** [comment]
GTX Titan on Windows and Tesla K40 on Ubuntu.

**Comment by pavanky** [comment]
@villekf any chance you can try to send some reproducible code ?

**Comment by villekf** [comment]
This kind of code seems to reproduce it:

    float indeksi[5]={1,2,3,4,5};
    int row[2] = {0, 5};
    int col[5] = {0, 10 , 100, 1000, 4000};
    array S1 = sparse(4096, 4096, 5, indeksi, row, col, f32, AF_STORAGE_CSR, afHost);
    output = matmul(S1,constant(1,4096,1));

As a side-note, sparse matrix created from ArrayFire array seems to work fine in matmul.

**Comment by pavanky** [comment]
@villekf for CSR, the row should be the size of the actual row + 1. In your case it should be 4097.

This works:
```
        float indeksi[5]={1,2,3,4,5};
        int row[4097] = {0};
        for (int i = 0; i < 4096; i++) row[i + 1] = 5;
        int col[5] = {0, 10 , 100, 1000, 4000};
        array r(2, row);
        array c(5, col);
        array val(5, indeksi);
        array S1 = sparse(4096, 4096, 5, indeksi, row, col, f32, AF_STORAGE_CSR, afHost);
        array D1 = af::randu(4096, 1);
        array output = matmul(S1, D1);
        af_print(output);
```
That said we should probably error out more gracefully when encountering such issues.

**Comment by villekf** [comment]
Works fine when using randu, but with 

	array D1 = af::constant(1,4096, 1);

I get the same problem. Also, my actual code has row number + 1.

**Comment by pavanky** [comment]
@villekf Your original code has this:
```
float indeksi[5]={1,2,3,4,5};
int row[2] = {0, 5};
```
This is not row number + 1. I'll double check with af::constant

**Comment by pavanky** [comment]
@villekf Nvm, I see you said actual code not the bug report.

Anyway, af::constant seems to cause a stall in CUDA backend while OpenCL and CPU backend work. Can you check if other backends are working for you?

**Comment by villekf** [comment]
Both OpenCL and CPU backends seem to work. Need to run some more tests tomorrow.

**Comment by villekf** [comment]
After some further testing these two test codes do not work on CUDA backend when using sparse matrix (non-sparse is fine):

	int pikselit = 4096;
	float nf = 7e-5f;
	float T = 0.5f;

	array G_ff = identity(pikselit * 2, pikselit)*nf;

	array F = sparse(join(1, identity(pikselit, pikselit), identity(pikselit, pikselit)*T));

	G_ff = matmul(F, G_ff);

and

	array S1 = sparse(identity(pikselit, pikselit));
	array X = randn(pikselit, 10);

	array output = matmul(S1, mean(X,1));

The first case works with a sparse matrix if *nf is removed from G_ff. Both of these work on OpenCL backend. However, this code works on CUDA backend, but not on OpenCL (when using sparse matrix):

	const array Cw = sparse(identity(pikselit * 2, pikselit * 2));
	array D = identity(pikselit * 2, pikselit * 2);
	D = matmul(D, matmul(Cw, transpose(D)));

The error message is:

	Unexpected Standard exception from MEX file.
	What() is:ArrayFire Exception (Internal error:998):
	In function class opencl::Array<float> __cdecl opencl::matmul<float>(const class opencl::Array<float>
	&,const class opencl::Array<float> &,af_mat_prop,af_mat_prop)
	In file src\backend\opencl\blas.cpp:103
	clblas Error (-48): Unknown error

	In function class af::array __cdecl af::matmul(const class af::array &,const class af::array &,const
	af_mat_prop,const af_mat_prop)
	In file src\api\cpp\blas.cpp:20

**Comment by pavanky** [comment]
Fixed in hotfixes-3.5.1
